src/SuperTokens.AspNetCore/SuperTokensHandler.cs

```diff
@@ -148,28 +148,42 @@ protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
             }
 
             var handshake = await _handshakeContainer.GetHandshakeAsync(this.Options.CoreApiKey, null, this.Context.RequestAborted);
-            AccessToken? parsedAccessToken;
+            AccessToken? parsedAccessToken = null;
+
+            // If we have no key old enough to verify this access token we should reject it without calling the core
+            var foundSigningKeyOlderThanToken = false;
             var isSignatureValid = false;
-            if (handshake.JwtSigningPublicKeyExpiration > now)
+            foreach (var keyInfo in handshake.AccessTokenSigningPublicKeyList)
             {
-                if (!JwtUtilities.TryParseAndValidate(accessToken, handshake.JwtSigningPublicKey, out var jwtPayload, out isSignatureValid) ||
+                if (!JwtUtilities.TryParseAndValidate(accessToken, keyInfo.PublicKey, out var jwtPayload, out isSignatureValid) ||
```

Comment:
There's already a `TryParse` method, and the `TryParseAndValidate` method was supposed to be an optimized routine for this specific use case. Parsing in a loop invalidates the optimization, and it's possible the `TryParseAndValidate` method doesn't make sense anymore. Maybe it'd be better to parse it just once and then, as an example, loop over each of the signing keys calling a `TryValidate` method?

Reply:
I split it up into multiple steps and moved all parsing outside the loop.